xbmc/utils/log.cpp

```diff
@@ -9,208 +9,198 @@
 #include "log.h"
 
 #include "CompileInfo.h"
-#include "ServiceBroker.h"
-#include "settings/AdvancedSettings.h"
-#include "settings/SettingsComponent.h"
-#include "threads/CriticalSection.h"
-#include "threads/SingleLock.h"
-#include "threads/Thread.h"
-#include "utils/StringUtils.h"
-
-#include <inttypes.h>
-
-#if defined(TARGET_POSIX)
-#include "platform/posix/utils/PosixInterfaceForCLog.h"
-typedef class CPosixInterfaceForCLog PlatformInterfaceForCLog;
-#elif defined(TARGET_WINDOWS)
+#include "filesystem/File.h"
+#include "utils/URIUtils.h"
+
+#if defined(TARGET_ANDROID)
+#include "platform/android/utils/AndroidInterfaceForCLog.h"
+#elif defined(TARGET_DARWIN)
+#include "platform/darwin/utils/DarwinInterfaceForCLog.h"
+#elif defined(TARGET_WINDOWS) || defined(TARGET_WIN10)
 #include "platform/win32/utils/Win32InterfaceForCLog.h"
-typedef class CWin32InterfaceForCLog PlatformInterfaceForCLog;
+#else
+#include "platform/posix/utils/PosixInterfaceForCLog.h"
 #endif
 
+#include <cstring>
 
-static const char* const levelNames[] =
-{"DEBUG", "INFO", "NOTICE", "WARNING", "ERROR", "SEVERE", "FATAL", "NONE"};
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/sinks/dist_sink.h>
 
-// add 1 to level number to get index of name
-static const char* const logLevelNames[] =
-{ "LOG_LEVEL_NONE" /*-1*/, "LOG_LEVEL_NORMAL" /*0*/, "LOG_LEVEL_DEBUG" /*1*/, "LOG_LEVEL_DEBUG_FREEMEM" /*2*/ };
+static constexpr unsigned char Utf8Bom[3] = { 0xEF, 0xBB, 0xBF };
+static const std::string LogFileExtension = ".log";
+static const std::string LogPattern = "%T.%e T:%-5t %8l %n: %v";
 
-namespace
-{
-class CLogGlobals
-{
-public:
-  ~CLogGlobals() = default;
-  PlatformInterfaceForCLog m_platform;
-  int         m_repeatCount = 0;
-  int         m_repeatLogLevel = -1;
-  std::string m_repeatLine;
-  int         m_logLevel = LOG_LEVEL_DEBUG;
-  int         m_extraLogLevels = 0;
-  CCriticalSection critSec;
-};
-
-static CLogGlobals g_logState;
-}
+bool CLog::m_initialized = false;
+std::shared_ptr<spdlog::sinks::dist_sink_mt> CLog::m_sinks = nullptr;
+Logger CLog::m_defaultLogger = nullptr;
 
-CLog::CLog() = default;
+#if defined(TARGET_ANDROID)
```

Comment:
instead of all this ifdeffing could you not create a static function in the interface. Something like this?

IPlatformLog.h
```
static std::unique_ptr<IPlatformLog> IPlatformLog::CreatePlatformInterface()
```

Then in each interface define the function?
PosixInterfaceForCLog.cpp
```
std::unique_ptr<IPlatformLog> CPosixInterfaceForCLog::CreatePlatformInterface()
{
  return std::make_unique<CPosixInterfaceForCLog>();
}
```

Reply:
Either I don't understand what you mean or I don't see the benefit. At some point we will have to have `#ifdef`s to include the proper headers. I can only combine the includes with the instantiations to reduce the two `#ifdef` blocks to a single one. Having a static method in an interface doesn't make any sense.

Reply:
Sure, it doesn't have to exist in the interface, but a base class on top would be fine. See CCPUInfo for an example of what I mean: https://github.com/xbmc/xbmc/blob/master/xbmc/utils/CPUInfo.h#L76
https://github.com/xbmc/xbmc/blob/master/xbmc/platform/linux/CPUInfoLinux.cpp#L65

Reply:
Done, I've added `IPlatformLog::CreatePlatformLog()`.